---
title: "Barcode analysis for Single-cell copy number variant detection reveals the dynamics and diversity of adaptation"
author: "Grace Avecilla, Julie Chuong"
date: "8/1/2018, 1-12-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "julie_barseq_csv_outfiles") #setwd where the bartender csv outfiles are. 
```

## This describes the barcode lineage tracking analysis
If you have any questions about this code or output, please contact Grace Avecilla ga824@nyu.edu or David Gresham dgresham@nyu.edu

# Introduction to 12-15-20 Barseq Dataset (Julie)
PATH to our BARSEQ DATA: /scratch/cgsb/gencore/out/Gresham/2020-12-15_HNYY3DRXX/2
- Paired-end 150 (PE150) not SR. So we had to think about which reads to use, R1 or R2, and how to use them. 
- R1 reads are the forward reads that go in this direction: fw primer->BC2->loxP->BC1-->rev primer
- R2 reads are the reverse reads that go in this direction: rev primer->BC1->loxP->BC2->fw primer
- The barcode of interest in each cell is BC2. BC2 is the unique barcode sequence for that cell. BC1 sequence is constant in all cells so don't use BC1 for any analysis. BC1 is there the purposes of rebarcoding, but we did not do that. BC1 and BC2 refers to the annotation file/Snapgene file of a barseq library. 

# Bartender Package
https://github.com/LaoZZZZZ/bartender-1.1

```{How to use R1 and R2 reads if barseq was PE150}
#R1 files in the direction fw primer->BC2->loxP->BC1-->rev primer, use the following patterns:
# pattern TACC[5]AA[5]AA[5]TT[5]ATAA  
# ^pattern written out: TACCNNNNNAANNNNNAANNNNNTTNNNNNATAA to search for BC2 in R1 files
# pattern containing BC1 (it is NOT the reverse complement of BC2!) : TTATNNNNNAANNNNNAANNNNNTTNNNNNGGTA to search for BC1 in R1 files...full pattern is truncated in the READ. *can do an ifelse statements to let it miss 2 bases from the end, 3bp, 4bp, 5bp etc to a cutoff we are comfortable with. #reverse comp is TTATNNNNNAANNNNNTTNNNNNTTNNNNNGGTA

#R2 files in the direction rev primer->BC1->loxP->BC2->fw primer
# pattern_withBC2(reverse comp of R1 BC2 pattern): TTATNNNNNAANNNNNTTNNNNNTTNNNNNGGTA  #use this pattern to search for BC2 sequence in R2 files. BC2 sequence is NOT truncated in the read! 

# Options/my thoughts for workflow. 
# 1. Do R1 only. Search for the BC2-containing pattern using bartender.  
# 2. Use R2 only. Search for reverse complement of the BC2-containing pattern from 1, using bartender.  
# 3. Number of unique barcodes found should be the same.

```


```{bash Barcode extraction and clustering}
# Run this bash script on the terminal in HPC
# USAGE: sbatch barcode_extract_and_cluster.sh 

# We had 12 fastq's, so set the array 0-11, one array per fastq

#!/bin/bash                     
#SBATCH --job-name=extract_and_cluster
#SBATCH --time=01:00:00
#SBATCH --nodes=1
#SBATCH --cpus-per-task=1
#SBATCH --array=0-11
#SBATCH --mem=8GB

## divide all fastq into R1 and R2
R1=($(ls /scratch/cgsb/gencore/out/Gresham/2020-12-15_HNYY3DRXX/2/*n01*.fastq.gz)) 
R2=($(ls /scratch/cgsb/gencore/out/Gresham/2020-12-15_HNYY3DRXX/2/*n02*.fastq.gz)) 

##this is going to assign the variables to file names
F_ID_R1=${R1[$SLURM_ARRAY_TASK_ID]}
F_ID_R2=${R2[$SLURM_ARRAY_TASK_ID]}

NAME=${F_ID_R1:75:-9} ##sample name, eg. bc02_g124

## make directories
mkdir ${NAME} 

## handling R1 F reads only from this point on
cp $F_ID_R1 $NAME/ ##copy each fastq to its newly made {NAME} directory
gunzip -c $NAME/*.gz > $NAME/"${NAME}_R1.fastq" ##unzip all the gz files inside the NAME directory, output the result into a new file named "{NAME}_R1.fastq" stored within the NAME dir 

## load bartender package
module purge
module load bartender/intel/1.1-20210106

##forward strand "BC2" extraction with quality cutoff of 30
bartender_extractor_com -f $NAME/${NAME}_R1.fastq -o $NAME/miniBar_F_$NAME -p TACC[5]AA[5]AA[5]TT[5]ATAA -u 0,8 -d f -q ?

## clustering (combining off-by-1 barcodes with exact match barcodes)
## use this line of code for cases in which we did not expect low coverage 
bartender_single_com -f $NAME/miniBar_F_${NAME}_barcode.txt -o F_$NAME -z -1  ##-z-1 denotes off-by-1

```

